File: bootstrap_for_pwm.jl
import CSV
import DataFrames
import Random
using ArgParse

function read_sites(path)
    container = Array{Char, 1}[]
    for line in eachline(path)
        if '>' != line[1]
            container = push!(container, collect(line))
        else
            continue
        end
    end
    return(container)
end


function create_pcm(sites)
    l = length(sites[1])
    pcm = Dict{Char, Array{Float64, 1}}(
        'A' => zeros(Float64, l),
        'C' => zeros(Float64, l),
        'G' => zeros(Float64, l),
        'T' => zeros(Float64, l))

    for i in sites, (index, nuc) in enumerate(i)
        pcm[nuc][index] += 1.0
    end
    return(pcm)
end


function create_pfm(pcm::Dict{Char,Array{Float64, 1}}, number_of_sites::Int64)
    pfm = Dict{Char, Array{Float64, 1}}()
    for nuc in keys(pcm)
        pfm[nuc] = (pcm[nuc] .+ 1.0/number_of_sites) ./ (number_of_sites + 1.0)
    end
    return(pfm)
end


function create_pwm(pfm::Dict{Char,Array{Float64, 1}}, background::Dict{Char,Float64})
    pwm = Dict{Char, Array{Float64, 1}}()
    for nuc in keys(pfm)
        pwm[nuc] = log2.(pfm[nuc] ./ background[nuc])
    end
    return(pwm)
end


function calculate_background(sites)
    all_nucleotides = join(sites)
    l = length(all_nucleotides)
    background = Dict{Char, Float64}()
    for nuc in all_nucleotides
        if !(nuc in keys(background))
            background[nuc] = 1
        else
            background[nuc] += 1
        end
    end

    for nuc in keys(background)
        background[nuc] = background[nuc] / l
    end
    return(background)
end


function make_pwm(sites)
    number_of_sites = length(sites)
    pcm = create_pcm(sites)
    pfm = create_pfm(pcm, number_of_sites)
    background = calculate_background(sites)
    pwm = create_pwm(pfm, background)
    return(pwm::Dict{Char,Array{Float64, 1}})
end


function calculate_score(site, pwm::Dict{Char,Array{Float64, 1}})
    score = 0.0
    for (index, nuc) in enumerate(site)
        score += pwm[nuc][index]
    end
    return(score)
end


function calculate_scores(sites, pwm::Dict{Char,Array{Float64, 1}})
    scores = Float64[]
    for site in sites
        score = calculate_score(site, pwm)
        scores = push!(scores, score)
    end
    return(scores)
end


function bootstrap_pwm(sites, size_of::Int64)
    true_scores = Float64[]
    false_scores = Float64[]
    number_of_sites = length(sites)

    for i in 1:10
        index_train = Random.randsubseq(1:number_of_sites, 0.9)
        index_test = setdiff(1:number_of_sites, index_train)
        index_shuffle = Random.rand(1:number_of_sites, size_of)

        pwm = make_pwm(sites[index_train])
        true_scores = vcat(true_scores, calculate_scores(sites[index_test], pwm))
        false_scores = vcat(false_scores, calculate_scores(
                Random.shuffle.(sites[index_shuffle]),
                pwm))
    end

    println(typeof(false_scores[1]))

    true_scores = sort(true_scores, rev=true)
    false_scores = sort(false_scores, rev=true)

    tpr = Float64[]
    tpr_actual = Float64[]
    fpr = Float64[]
    scores = Float64[]

    false_length = length(false_scores)
    true_length = length(true_scores)

    for i in 0.05:0.05:1.0
        s = true_scores[Int(round(true_length * i))]
        tpr_actual = push!(tpr_actual, sum(true_scores .>= s) / true_length)
        tpr = push!(tpr, i)
        fpr = push!(fpr, sum(false_scores .>= s) / false_length)
        scores = push!(scores, s)
    end

    df = DataFrames.DataFrame(Scores = scores, TPR = tpr, ACTUAL_TPR = tpr_actual, FPR = fpr)

    return(df)
end


function parse_commandline()
    s = ArgParseSettings()

    @add_arg_table s begin
        "output"
            help = "path to write results"
            required = true
        "input"
            help = "path to sites of PWM model"
            required = true
        "--size", "-s"
            help = "size of negative sites on each bootstrap iteration"
            arg_type = Int
            default = 100000

    end
    return parse_args(s)
end


function main()
    args = parse_commandline()
    path = args["input"]
    out = args["output"]
    size_of = args["size"]
    sites = read_sites(path)
    df = bootstrap_pwm(sites, size_of)
    CSV.write(out, df, delim='\t')
end

# function main()
#     s = parse_commandline()
#     println("Parsed args:")
#     println(s)
#     println(s["input"])
#     println(s["output"])
#     println(s["size"])
#     for (arg,val) in s
#         println("  $arg  =>  $val")
#     end
# end

main()

File: bootstrap_for_pwm.py
import csv
import itertools
import os
import sys
import math
from math import log
import re
import random
import subprocess
import shutil
import time
import argparse


def make_pcm(motifs):
    matrix = {}
    mono_nucleotides = itertools.product('ACGT', repeat=1)
    for i in mono_nucleotides:
        matrix[i[0]] = []
    len_of_motif = len(motifs[0])
    for i in matrix.keys():
        matrix[i] = [0]*len_of_motif
    for i in range(len_of_motif):
        for l in motifs:
            matrix[l[i]][i] += 1
    return(matrix)


def make_pfm(pcm):
    number_of_sites = [0] * len(pcm['A'])
    for key in pcm.keys():
        for i in range(len(pcm[key])):
            number_of_sites[i] += pcm[key][i]
    pfm = dict()
    mono_nucleotides = itertools.product('ACGT', repeat=1)
    for i in mono_nucleotides:
        pfm[i[0]] = []
    first_key = list(pcm.keys())[0]
    nuc_pseudo = 1/len(pcm.keys())
    for i in range(len(pcm[first_key])):
        for nuc in pcm.keys():
            pfm[nuc].append((pcm[nuc][i] + nuc_pseudo) / (number_of_sites[i] + 1))
    return(pfm)


def make_pwm(pfm):
    pwm = {}
    background = {'A': 0.25,
                 'C': 0.25,
                 'G': 0.25,
                 'T': 0.25}
    mono_nucleotides = itertools.product('ACGT', repeat=1)
    for i in mono_nucleotides:
        pwm[i[0]] = []
    first_key = list(pfm.keys())[0]
    for i in range(len(pfm[first_key])):
        for j in pfm.keys():
            pwm[j].append(math.log(pfm[j][i] / background[j]))
    return(pwm)


def sites_to_pwm(sites):
    pcm = make_pcm(sites)
    pfm = make_pfm(pcm)
    pwm = make_pwm(pfm)
    return(pwm)


def parse_chipmunk(path):
    with open(path, 'r') as file:
        container = []
        for line in file:
            d = {'name': str(), 'start': int(), 'end': int(),
                 'seq': str(), 'strand': str()}
            if line.startswith('WORD|'):
                line = line[5:].strip()
                line = line.split()
                d['name'] = 'peaks_' + str(int(line[0]) - 1)
                d['start'] = int(line[1])
                d['end'] = int(line[1]) + len(line[2])
                d['seq'] = line[2]
                if line[4] == 'direct':
                    d['strand'] = '+'
                else:
                    d['strand'] = '-'
                container.append(d)
            else:
                continue
    seqs = [i['seq'] for i in container if not 'N' in i['seq']]
    return(seqs)


def complement(seq):
    return(seq.replace('A', 't').replace('T', 'a').replace('C', 'g').replace('G', 'c').upper()[::-1])


def run_chipmunk(path_to_java, path_to_chipmunk, fasta_path, path_out, motif_length_start, motif_length_end, cpu_count):
    args = [path_to_java, '-cp', path_to_chipmunk,
                   'ru.autosome.ChIPMunk', str(motif_length_start), str(motif_length_end), 'yes', '1.0',
                   's:{}'.format(fasta_path),
                  '100', '10', '1', str(cpu_count), 'random']
    p = subprocess.run(args, shell=False, capture_output=True)
    out = p.stdout
    with open(path_out, 'wb') as file:
        file.write(out)
    return(0)


def write_fasta(peaks, path):
    with open(path, 'w') as file:
        for index, p in enumerate(peaks):
            file.write('>{}\n'.format(index))
            file.write(p + '\n')
    return(0)


def read_peaks(path):
    container = []
    append = container.append
    with open(path) as file:
        for line in file:
            if not line.startswith('>'):
                append(line.strip().upper())
    return(container)


def score_pwm(seq, pwm):
    score = 0
    position = 0 
    length = len(seq)
    for index in range(length):
        score += pwm[seq[index]][position]
        position += 1
    return(score)


def false_scores_pwm(peaks, pwm):
    false_scores = []
    length_pwm = len(pwm['A'])
    append = false_scores.append
    for peak in peaks:
        complement_peak = complement(peak)
        length_pwm = len(pwm['A'])
        full_peak = peak + 'N' * length_pwm + complement_peak
        n = len(full_peak) - length_pwm + 1
        for i in range(n):
            site = peak[i:length_pwm + i]
            if 'N' in site:
                continue
            score = score_pwm(site, pwm)
            false_scores.append(score)
    return(false_scores)


def true_scores_pwm(peaks, pwm):
    true_scores = []
    for peak in peaks:
        complement_peak = complement(peak)
        length_pwm = len(pwm['A'])
        best = -1000000
        full_peak = peak + 'N' * length_pwm + complement_peak
        n = len(full_peak) - length_pwm + 1
        for i in range(n):
            site = full_peak[i:length_pwm + i]
            if 'N' in site:
                continue
            score = score_pwm(site, pwm)
            if score >= best:
                best = score
        true_scores.append(best)
    return(true_scores)


def creat_background(peaks, length_of_site, counter):
    shuffled_peaks = []
    number_of_sites = 0
    while counter > number_of_sites:
        peak = random.choice(peaks)
        shuffled_peak = ''.join(random.sample(peak, len(peak)))
        shuffled_peaks.append(shuffled_peak)
        number_of_sites += (len(''.join(shuffled_peak)) - length_of_site + 1) * 2
    return(shuffled_peaks)


def bootstrap_pwm(peaks, length_of_site, counter, path_to_java, path_to_chipmunk, tmp_dir, cpu_count):
    true_scores = []
    false_scores = []
    number_of_peaks = len(peaks)
    for i in range(5):
        train_peaks = random.choices(peaks, k=round(0.9 * number_of_peaks))
        test_peaks = [peak for peak in peaks if not peak in train_peaks]
        shuffled_peaks = creat_background(test_peaks, length_of_site, counter / 5)
        write_fasta(train_peaks, tmp_dir + '/train.fasta')
        run_chipmunk(path_to_java, path_to_chipmunk,
                     tmp_dir + '/train.fasta', tmp_dir + '/chipmunk_results.txt',
                     length_of_site, length_of_site, cpu_count)
        sites = parse_chipmunk(tmp_dir + '/chipmunk_results.txt')
        sites = list(set(sites))
        nsites = len(sites)
        pcm = make_pcm(sites)
        pfm = make_pfm(pcm)
        pwm = make_pwm(pfm)
        for true_score in true_scores_pwm(test_peaks, pwm):
            true_scores.append(true_score)
        for false_score in false_scores_pwm(shuffled_peaks, pwm):
            false_scores.append(false_score)
    table = creat_table_bootstrap(true_scores, false_scores)
    return(table)


def creat_table_bootstrap(true_scores, false_scores):
    table = []
    true_scores.sort(reverse=True)
    false_scores.sort(reverse=True)
    false_length = len(false_scores)
    true_length = len(true_scores)
    for tpr in [round(i * 0.01, 2) for i in range(5,105, 5)]:
        score = true_scores[round(true_length * tpr) - 1]
        actual_tpr = sum([1 if true_score >= score else 0 for true_score in true_scores]) / true_length
        fpr = sum([1 if false_score >= score else 0 for false_score in false_scores]) / false_length
        table.append({'Scores': score, 'TPR': tpr, 'ACTUAL_TPR': actual_tpr, 'FPR': fpr})
    return(table)


def write_table_bootstrap(path, data):
    with open(path, 'w') as csvfile:
        fieldnames = data[0].keys()
        writer = csv.DictWriter(csvfile, fieldnames=fieldnames, delimiter='\t')
        writer.writeheader()
        for line in data:
            writer.writerow(line)
    return(0)

def parse_args():
    parser = argparse.ArgumentParser()
    parser.add_argument('fasta', action='store', help='path to file with peaks')
    parser.add_argument('results', action='store', help='path to write table with ROC')
    parser.add_argument('length', action='store', type=int, help='length of TFBS')
    parser.add_argument('chipmunk', action='store', help='path to ChIPMunk source')
    parser.add_argument('-j', '--java', action='store', type=str, dest='java',
                        required=False, default='java', help='Path to java')
    parser.add_argument('-t', '--tmp', action='store', type=str, dest='tmp',
                        required=False, default='./pwm.tmp', help='tmp directory')
    parser.add_argument('-p', '--processes', action='store', type=int, dest='cpu_count',
                        required=False, default=4, help='Number of processes to use, default: 4')
    if len(sys.argv) == 1:
        parser.print_help(sys.stderr)
        sys.exit(1)
    return(parser.parse_args())
    
    
def main():
    args = parse_args()
    peaks_path = args.fasta
    results_path = args.results
    length_of_site = args.length
    path_to_chipmunk = args.chipmunk
    path_to_java = args.java
    tmp_dir = args.tmp
    cpu_count = args.cpu_count
    if not os.path.exists(tmp_dir):
        os.mkdir(tmp_dir)        
    counter = 5000000
    peaks = read_peaks(peaks_path)
    table = bootstrap_pwm(peaks, length_of_site, counter, path_to_java, path_to_chipmunk, tmp_dir, cpu_count)
    write_table_bootstrap(results_path, table)
    shutil.rmtree(tmp_dir)
    return(0)


def bootstrap_for_pwm(peaks_path, results_path, length_of_site, path_to_java, path_to_chipmunk, tmp_dir, cpu_count, counter=5000000):
    if not os.path.exists(tmp_dir):
        os.mkdir(tmp_dir)
    peaks = read_peaks(peaks_path)
    table = bootstrap_pwm(peaks, length_of_site, counter, path_to_java, path_to_chipmunk, tmp_dir, cpu_count)
    write_table_bootstrap(results_path, table)
    shutil.rmtree(tmp_dir)
    return(0)


if __name__=="__main__":
    main()

